sound/soc/sof/sof-priv.h

```
@@ -148,6 +148,16 @@ struct snd_sof_dsp_ops {
 	snd_pcm_uframes_t (*pcm_pointer)(struct snd_sof_dev *sdev,
 					 struct snd_pcm_substream *substream); /* optional */
 
+	/* host read DSP stream data */
+	void (*ipc_msg_data)(struct snd_sof_dev *sdev,
+			     struct snd_pcm_substream *substream,
+			     void *p, size_t sz);
+
+	/* host configure DSP HW parameters */
+	int (*ipc_pcm_params)(struct snd_sof_dev *sdev,
+			      struct snd_pcm_substream *substream,
+			      const struct sof_ipc_pcm_params_reply *reply);
```

add mandatory/optional tags for the two functions?

---

TBH, I find the use of those tags not perfectly consistent in SOF. E.g. .irq_handler and .irq_thread are marked as mandatory and hsw.c doesn't have them. In general, I would check for all mandatory ops at the initialisation time and fail if any of them is missing, rather than printing error messages at run-time.

---

ok, then please report/fix those as well :-)

---

ok, I'll make an additional PR